pkg/es/config/config_test.go

```diff
@@ -0,0 +1,740 @@
+// Copyright (c) 2024 The Jaeger Authors.
+//
+
+package config
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+
+	"github.com/jaegertracing/jaeger/pkg/config/tlscfg"
+	"github.com/jaegertracing/jaeger/pkg/metrics"
+)
+
+var mockEsServerResponseWithVersion0 = []byte(`
+{
+	"Version": {
+		"Number": "0"
+	}
+}
+`)
+
+var mockEsServerResponseWithVersion1 = []byte(`
+{
+	"tagline": "OpenSearch",
+	"Version": {
+		"Number": "1"
+	}
+}
+`)
+
+var mockEsServerResponseWithVersion2 = []byte(`
+{
+	"tagline": "OpenSearch",
+	"Version": {
+		"Number": "2"
+	}
+}
+`)
+
+func TestNewClient(t *testing.T) {
+	const (
+		pwd1        = "password"
+		token       = "token"
+		certContent = `-----BEGIN CERTIFICATE-----
+MIIDMTCCAhkCFCi65dSe1JONpNGghyam61+4gTL7MA0GCSqGSIb3DQEBCwUAMFUx
+CzAJBgNVBAYTAkFVMRIwEAYDVQQIDAlBdXN0cmFsaWExDzANBgNVBAcMBlN5ZG5l
+eTEQMA4GA1UECgwHTG9nei5pbzEPMA0GA1UEAwwGSmFlZ2VyMB4XDTIyMDkxMDAw
+MjE0NFoXDTMyMDkwNzAwMjE0NFowVTELMAkGA1UEBhMCQVUxEjAQBgNVBAgMCUF1
+c3RyYWxpYTEPMA0GA1UEBwwGU3lkbmV5MRAwDgYDVQQKDAdMb2d6LmlvMQ8wDQYD
+VQQDDAZKYWVnZXIwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDLlEq/
+DF2pkhfSedvAd5h6BXCjpC/mUA6BN3RyMHUjTWr9hhBtaIYv68O12GMVf//ST/Fs
+CjRrjOcqrz2QQn3P8UelGRd2vJfcMhJElQ/lnKmZZlAHEOMF8TC7nQfsReLCwcpj
+T6bXqvDcfHjDye+45F2rPDpRGLzyysg7pgdINp0Duph0Z16ggrBgz7RVNBmWsYVe
+sGD3VOR3hLd8GTDzJ5amRpkq8nfliJ+U3JLGcDG/7Wkuvl/YZZxf21v9f4yYVEZZ
+aLAcKsHIUoFRDJtdrBeaPZRJjL/I9B1M6En+Styxb5wJw42h9BXtJd2IeQPp15pP
+KfPbkmOj+X+2s9n1AgMBAAEwDQYJKoZIhvcNAQELBQADggEBAJbm7WXgQirWQbaa
+E304K8tvdpC2E1ewxTTrUEN8jUONER4KC+epRnsTgkEpVlj7sehiAgSMnbT4E3ve
+GjmsUrZiJcKPaf+ogn49Cj0weD99wbJtUNgbH4HiqR1ePOHIRDQ7GD5G0zdFq7oO
+Il09eHAbbWM61x04I3XDQ0OwXyeVXIEWJcR1R6wnuNMJm54czbXvn6SrIuoMCvs6
+oSkVm43Q+plk0hlDZnA/KiOxqFRLVHBuX/SgRf5NBg8m7id3fNzIJnWWK+zqoDoZ
+ryja7dFIJnLqEXJxJkc5ubT1/j9PDE51WbM5MyPB6lnuQKdZTbDziyKiVXg0au3E
+QK5K/Ow=
+-----END CERTIFICATE-----`
+	)
+	pwdFile := filepath.Join(t.TempDir(), "pwd")
+	require.NoError(t, os.WriteFile(pwdFile, []byte(pwd1), 0o600))
+	pwdtokenFile := filepath.Join(t.TempDir(), "token")
+	require.NoError(t, os.WriteFile(pwdtokenFile, []byte(token), 0o600))
+	certFilePath := filepath.Join(t.TempDir(), "CA-cert.pem")
+	require.NoError(t, os.WriteFile(certFilePath, []byte(certContent), 0o644))
+	tests := []struct {
+		name          string
+		setup         func() (*Configuration, *zap.Logger, metrics.Factory)
+		expectedError bool
+	}{
+		{
+			name: "success with valid configuration",
+			setup: func() (*Configuration, *zap.Logger, metrics.Factory) {
```

Comment:
this doesn't pass the DRY smell test. Why do you need such extensive setup functions for each test case? What is different across them? If they are only difference in a few parameters, then define the parameters in the test case and move setup into the body of the test.

Reply:
yes this is much cleaner and clearer now : D